### PR TITLE
[magnum-auto-healer] Fix k8s master node repair if it's volume-based

### LINKED ISSUE #1446: [magnum-auto-healer] Magnum auto healer cannot handle volume based master nodes
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
When the k8s nodes are created based on volume, the repair will fail with the error that the root volume is still in in-use status.

**What you expected to happen**:
The volume-based nodes can be repaired successfully

**How to reproduce it**:
Create a k8s cluster with Magnum based on volume-based nodes, then stop a kubelet process in a master node, then magnum auto healer will try to fix it and fail.

**Anything else we need to know?**:

**Environment**:
- magnum-auto-healer version: v1.19.4
- OpenStack version: stable/victoria
- Others:

### PR DESCRIPTION
**Pull request by openstacker**
**What this PR does / why we need it**:
Fix k8s master node auto repair if the node is volume-based.

**Which issue this PR fixes(if applicable)**:
fixes #1446 

**Special notes for reviewers**:

**Release note**:
```release-note
 [magnum-auto-healer] Fixed an issue that k8s master node repair failed when the node is boot from volume.
```

### CODE REVIEW

**Comment by jichenjc** [comment]
@openstacker maybe you want to update your release note :)

**Review by lingxiankong — CHANGES_REQUESTED** [review]

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 1]
```diff
@@ -209,20 +232,26 @@ func (provider OpenStackCloudProvider) waitForServerDetachVolumes(serverID strin
...
-// the VMs. The information this function needs:
-//     - Nova VM IDs
-// 	   - Heat stack ID and resource ID.
+// Repair  For master nodes: detach etcd and docker volumes, find the root
+//         volume, then shutdown the VM, marks the both the VM and the root
+//         volume (heat resource) as "unhealthy" then trigger Heat stack update
+//         in order to rebuild the node. The information this function needs:
+//         - Nova VM ID
+//         - Root volume ID
+// 	       - Heat stack ID and resource ID.
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:254]: Resize can handle the volume-based node.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 2]
```diff
@@ -307,24 +356,54 @@ func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) erro
...
 				log.Warningf("Failed to shutdown the server %s, error: %v", serverID, err)
 			}
 
 			log.Infof("Marking Nova VM %s(Heat resource %s) unhealthy for Heat stack %s", serverID, allMapping[serverID].ResourceID, cluster.StackID)
 
-			opts := stackresources.MarkUnhealthyOpts{
-				MarkUnhealthy:        true,
-				ResourceStatusReason: "Mark resource unhealthy by autohealing service",
-			}
+			// Mark VM as unhealthy
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:406]: Yep, nice catch.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 3]
```diff
@@ -260,7 +290,25 @@ func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) erro
 			}
 			serverID := machineID.String()
 
-			if err := provider.waitForServerDetachVolumes(serverID, 30*time.Second); err != nil {
+			var firsttimeUnhealthy = true
+			for id := range unHealthyNodes {
+				log.Infof("comparing server ID %s with known broken ID %s", serverID, id)
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:295]: Will do.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 4]
```diff
@@ -307,24 +356,54 @@ func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) erro
...
+					log.Warningf("failed to reboot node %s, error: %v", serverID, err)
+				}
+				continue
+			}
+
+			if rootVolumeID, err := provider.waitForServerDetachVolumes(serverID, 30*time.Second); err != nil {
+				log.Warningf("Failed to detach volumes from server %s, error: %v", serverID, err)
+			} else {
+				// Mark root volume as unhealthy
+				err = stackresources.MarkUnhealthy(provider.Heat, allMapping[serverID].StackName, allMapping[serverID].StackID, rootVolumeID, opts).ExtractErr()
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:394]: I will add a IF check

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 5]
```diff
@@ -307,24 +356,54 @@ func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) erro
...
+					firsttimeUnhealthy = false
+					break
+				}
+			}
+
+			if firsttimeUnhealthy == true {
+				unHealthyNodes[serverID] = n
+				log.Infof("rebooting node %s to repair it", serverID)
+				if res := servers.Reboot(provider.Nova, serverID, servers.RebootOpts{Type: servers.SoftReboot}); res.Err != nil {
+					log.Warningf("failed to reboot node %s, error: %v", serverID, err)
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:385]: Nice catch. Will do.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 6]
```diff
@@ -260,7 +290,25 @@ func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) erro
...
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:306]: Reboot and resize are separated actions for the repairing. When there are more than one node failed, they're repaired separately.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 7]
```diff
@@ -184,23 +191,39 @@ func (provider OpenStackCloudProvider) waitForClusterComplete(clusterID string,
...
-			log.Infof("detaching volume %s for instance %s", attachment.VolumeID, serverID)
-			err := volumeattach.Delete(provider.Nova, serverID, attachment.ID).ExtractErr()
+			volume, err := volumes.Get(provider.Cinder, attachment.VolumeID).Extract()
 			if err != nil {
-				return false, fmt.Errorf("failed to detach volume %s from instance %s", attachment.VolumeID, serverID)
+				return false, fmt.Errorf("failed to get volume %s", attachment.VolumeID)
+			}
+
+			bootable, err := strconv.ParseBool(volume.Bootable)
+			log.Infof("volume %s is bootable %t", attachment.VolumeID, bootable)
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:208]: Good point.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 8]
```diff
@@ -184,23 +191,39 @@ func (provider OpenStackCloudProvider) waitForClusterComplete(clusterID string,
...
+			if err != nil {
+				log.Warningf("Unexpected value for bootable volume %s in volume %s", volume.Bootable, volume)
+			}
+			if bootable == false {
+				log.Infof("detaching volume %s for instance %s", attachment.VolumeID, serverID)
+				err := volumeattach.Delete(provider.Nova, serverID, attachment.ID).ExtractErr()
+				if err != nil {
+					return false, fmt.Errorf("failed to detach volume %s from instance %s", attachment.VolumeID, serverID)
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:216]: will do.

File: pkg/autohealing/cloudprovider/openstack/provider.go [hunk 9]
```diff
@@ -184,23 +191,39 @@ func (provider OpenStackCloudProvider) waitForClusterComplete(clusterID string,
 	return err
 }
 
-func (provider OpenStackCloudProvider) waitForServerDetachVolumes(serverID string, timeout time.Duration) error {
+func (provider OpenStackCloudProvider) waitForServerDetachVolumes(serverID string, timeout time.Duration) (string, error) {
```
[the remark this answers is not in the document]
Comment (openstacker) [reply pkg/autohealing/cloudprovider/openstack/provider.go:194]: Will do.

**Pull request merged**